Add test coverage to xk6-redis

oleiade opened the pull request:
## What

This PR adds integration tests covering the xk6-redis API with the goal to improve its stability and test coverage in preparation of merging it to k6's core. It attempts to test the extension in the context of a script interacting with a Redis instance, and focuses on testing the k6/goja/promises/event loop behavior of the `Client`, rather than interactions with Redis itself (considering we use a pre-established Redis library under the hood). In certain scenarios, we relied on a more extensive testing strategy in order to explicit the promise-based specific behavior of certain operations that might derive from Redis' behavior itself (rejecting a promise under certain conditions rather than returning zero, for instance).

Things that were left out of the scope of this PR: this PR doesn't test Redis in sentinel mode, nor in cluster mode. The feature should be supported out of the box, but I didn't get to address testing it, yet (if ever). 

## How

This PR is quite massive. Writing tests for an API as big as Redis' involves, well, lots of code. That was to be expected. However, the test file is organized as follows:
- Every `Client` command has its own `TestClient{Command}(` test function. They all follow the same convention, and the same approach, and should be somewhat similar. It's probably not worth reviewing ALL of them, but by reviewing a few of them, the pattern should become explicit, and you will likely be able to spot systemic mistakes and cases I might have missed. 
- Two table tests `TestClientCommandsAgainstUnreachableServer` and `TestClientCommandInInitContext` verify that, based on the fact that every command will ensure a connection is established under the hood, those should fail when ran against an unreachable host, or when executed in the init context.
- `Client.Connect`, `Client.Close` and private methods have been tested separately, and follow a bit of a different testing convention, and it might be reviewing those.

## Open questions and Feedback needed

### Connect and Close methods

Although the underlying [library](https://github.com/go-redis/redis) we used to implement the interactions with Redis does not require to explicitly establish a connection, we need to emulate that in the context of k6. The main reason for is that we want to avoid IO to happen in the init context of our users scripts. Besides, as we want our Redis client to reuse our VUs `Dialer` and `TLSConfig`, we need to wait for the VU `State` to be available before any connection to a Redis instance/cluster is attempted. As a result, I tackled this constraint by implementing explicit `Connect` and `Close` methods, as we seem to already have in other extensions/modules such as gRPC's. 

Our main design question is: as of today, every client's methods call connect under the hood, to make sure that the Redis client has an underlying client's instance, that is properly configured with the VU's `Dialer` and `TLSConfig`. I would like to hear your views on what design you'd prefer: **the user must explicitly call the `.connect` method in the init context, or we keep connecting under the hood automatically?**. 

### Testing approach

We've used an integration testing approach. To avoid importing extra-dependencies, we have developed a dedicated and externally controllable redis mock/stub server (in the spirit of `httptest`). Running this stub server, we run scripts directly into a test runtime, and assert the outcome directly in the script. Once the script is successfully run, we assert that our redis stub server received the commands we expected in the order we anticipated.

I probably have missed some failure scenarios, some edge cases, or dark corners of the script I'm not yet aware of, or have forgotten 🤷🏻 Please, check out some tests, and let me know if you can think of ways to break them, so I can improve the whole suite.

## Finally

Thank you for your time and attention! Please let me know what you think 🙇🏻 

## Edits

**26th of July 2022**: Removed mentions of miniredis in favor of the stub server. Explicated testing design strategy.

mstoykov commented on client.go:
```diff
@@ -1026,6 +1030,8 @@ func (c *Client) connect() error {
 		return common.NewInitContextError("connecting to a redis server in the init context is not supported")
 	}
 
+	// If the redisClient is already instantiated, it is safe
+	// to assume that the connection is already established.
```
Well we also can have the connection disconnected :shrug: But I would guess that will be rare and we can create an issue and fix it when we start seeing it.

mstoykov commented on client.go:
```diff
@@ -1060,7 +1095,7 @@ func (c *Client) isSupportedType(offset int, args ...interface{}) error {
 		switch arg.(type) {
 		case string, int, int64, float64, bool:
-			return nil
+			continue
```
nit: I kind of prefer the previous one :thinking:

oleiade replied:
But it doesn't actually work, hence the dedicated test I wrote illustrating it 👍🏻 
If you return, you do not handle all the args, thus if any of them is actually invalid (when handling arrays for instance), you miss it, but redis won't 🙇🏻

mstoykov replied:
aaah .... yeah, correct :facepalm:

mstoykov commented on redis.go:
```diff
@@ -1,4 +1,5 @@
 // Package redis implements a redis client for k6.
+// nolint: exhaustruct
```
I am pretty sure we will not enable this linter after all

also (and this will be enforced in go 1.19) comment directives (comments not for humans) are supposed to be without spaces between `//` and the namespace (`nolint` in this case) and no space around the `:` you can see https://github.com/golangci/golangci-lint/issues/1658 and the linked issues.

oleiade replied:
Thank you so much for the heads-up 🙇🏻 Will adjust accordingly, and remove the linter from the repo's config.

oleiade replied:
Fixed 👍🏻 Thanks for the heads-up 🙇🏻

mstoykov replied:
Either github is showing some cached version or it isn't fixed ;)

oleiade replied:
Damn, I missed that one, sorry for that 🙇🏻

mstoykov commented on go.mod:
```diff
@@ -3,26 +3,43 @@ module github.com/grafana/xk6-redis
 	github.com/serenize/snaker v0.0.0-20201027110005-a7ad2135616e // indirect
 	github.com/sirupsen/logrus v1.8.1 // indirect
 	github.com/spf13/afero v1.1.2 // indirect
+	github.com/yuin/gopher-lua v0.0.0-20210529063254-f4c35e4016d9 // indirect
```
:(

oleiade replied:
Gone 👍🏻

mstoykov commented on client_test.go:
```diff
@@ -0,0 +1,2068 @@
+	redis *miniredis.Miniredis
+}
+
+// newTestSetup initializes a new test setup.
+// It prepares a test setup with a mocked redis server and a goja runtime,
+// and event loop, ready to execute scripts as if being executed in the
+// main context of k6.
```
Haven't looked through all the code but in general it might be better to just have a method on testSetup to move from init context to vu context or somethign like that :shrug:

mstoykov reviewed (commented):
I haven't looked at the majority of the test code but would like to ask:

Have we considered not using a full blown redis in golang as this seems to pull a bunch of dependancies (even just for tests) that would then end up in the core k6 repo. 

Arguably we don't really care about the "correct" behaviuour of any of the commands once they are called. So I wonder if we can just have some stubs that just check that it gets what seems like an okay redis command. 

This arguably will require some amount of redis wire format parsing and returning of somethign that looks at least somewhat like what redis client should expect. 

Running a full blown redis and having the integration tests behind a flag might also  be preferable if it will mean that we will not suddenly depend on a lua interpreter for some tests :(

oleiade commented:
Changelog since the last review round:
- I have dropped the dependency on miniredis, and replaced it with a custom redis stub server, which listens on a TCP port, talks the redis SPEC protocol, is fully controllable, and reports stuff such has command called, command history, connections counts etc... It has zero external dependency and most importantly allows registering custom command handlers, which you can make reply, well, whatever you need your redis stub server to reply. 
- I have rewritten most tests to be dumbed down versions of themselves. They now for the most part, verify that given a bunch of redis commands and promises chains, the server receives expected correctly formatted messages, in the expected order. I kept certain test scenarios a bit richer, to explicit that although redis would not return an error in certain scenarios, our methods would sometimes reject promises.
- Tests have been rewritten to use promise chaining more efficiently. "Wow, I know JavaScript promises now"
-

mstoykov commented on stub.go:
```diff
@@ -0,0 +1,623 @@
+package redis
```
Can you move this to `stub_test.go` as it's only used in tests ;)

mstoykov reviewed (approved):
From a fairly quick over the top look - haven't looked at each test case separately through and through - it LGTM! :clap:

I would prefer to get the golangci.yaml update now and remove all the `nolint` that are possible to be removed. Especially file wide ones.

As general note on tests - it might be better to check the error - like just to be certain it doesn''t actually error with something we are not expecting.

It will be nice if we have the example from the Readme as  a test :thinking: Or at least something like it - that does something more. 

The stub looks a bit big and I did not have time to look it over all that much, sorry - maybe if this stays open longer I might go through it more.

You have a general :+1: from me, if you have something specific you want my opinion on - I can try to look into it more, but otherwise I don't see any reason to block merging this. We can always fix something that turned out to not be as a great down the line.

p.s. I thought about the fact that a lot of the tests almost look at the same and can be made into table ones ... but that likely will not make them any easier to read so ... I am kind of against that idea.

codebien commented on client.go:
```diff
@@ -1014,9 +1058,9 @@ func (c *Client) makeHandledPromise() (*goja.Promise, func(interface{}), func(in
+// Connect establishes the client's connection to the target
 // redis instance(s).
-func (c *Client) connect() error {
+func (c *Client) Connect() error {
```
>the user must explicitly call the .connect method in the init context, or we keep connecting under the hood automatically?

We went with the second so do we need to export `connect`? Do we expect a use case for it? If yes, I think we should add a test for it and maybe expand a bit the documentation here.

oleiade replied:
See me question in the PR description. The reason I export it is to be able to test it, as there are tests for it, at the end of the test file 😉 (that file is huge, I would have missed it myself).

The main rationale is: every method "connect" under the hood, to ensure that once in the VU context, the TLS config, and dialers are set up (it's forbidden to do that in the init context). To be able to test that connecting in the init context is forbidden, and that connecting to an unreachable server fails, I had to expose this method (so that I can test it on the JS side)

codebien replied:
>as there are tests for it, at the end of the test file wink (that file is huge, I would have missed it myself).
seen, thanks  :pray: 

>I had to expose this method (so that I can test it on the JS side)

It doesn't seem we really need the concrete JS runtime for those tests. The cases seem to me they are reproducible just using the `modulestest.NewRuntime(t)`. Am I wrong?

oleiade replied:
You're right! That's super useful input, I'm going to integrate it, thanks a bunch 🙇🏻

codebien commented on client.go:
```diff
@@ -1026,23 +1070,58 @@ func (c *Client) connect() error {
 }
 
+// Close closes the client's connection to the target redis instance(s).
+func (c *Client) Close() error {
```
Same as `Connect`

oleiade replied:
Same rationale as `Connect` 😄

oleiade commented on client_test.go:
```diff
@@ -0,0 +1,2644 @@
+//nolint:exhaustruct,goconst
```
Good point, gonna do that, and make sure to move the ones that are still needed to a case by case basis rather than file-wide 👍🏻

oleiade commented on stub.go:
```diff
@@ -0,0 +1,623 @@
+
+// newRedisServer creates a new redis server listening on the provided
+// network address.
+func newRedisServer(addr string) (*redisServer, error) {
```
Fair point, I considered it initially, but can't remember why I didn't make the move. Gonna reconsider 👍🏻

oleiade commented:
Alright, I've addressed the feedback I received, most notably:
- the file is now called `stub_test.go`
- the `redisServer` is now gone in favor of just `StubServer`
- the promise `err` that are expected to fail are now checked
- the `connect` and `close` methods are now private

codebien commented on stub_test.go:
```diff
@@ -0,0 +1,547 @@
+		rs.handleCommand(connection, command, args)
+		connection.Flush()
+
+		rs.Lock()
+		closed := connection.closed
+		rs.Unlock()
+		if closed {
+			nc.Close() //nolint:errcheck,gosec
+		}
```
the caller is passing an open connection so it's already managing `net.Conn`, if we can leave this responsibility centralized I think would be better. `listenAndServe` is already calling defer `nc.Close()` so I'm just for removing this block.

oleiade replied:
Good point indeed!

codebien commented on stub_test.go:
```diff
@@ -0,0 +1,540 @@
+	writer *bufio.Writer
+	reader *bufio.Reader
+	mutex  sync.Mutex
+	closed bool
```
```suggestion
```
It seems not used anymore so can we remove it?

codebien reviewed (approved):
LGTM :clap: :tada:
